gst/mqtt/ntputil.c

```diff
@@ -0,0 +1,289 @@
+        int64_t diff_usec = monotonic_usec - local_monotonic_usec;
+
+        if (diff_usec < ntp_poll_interval_usec) {
+          pthread_spin_unlock (&lock);
+          return (diff_usec + ntp_timestamp_usec);
+        }
+      }
+    }
+  }
+  pthread_spin_unlock (&lock);
+
+  sockfd = socket (AF_INET, SOCK_DGRAM, IPPROTO_UDP);
+  if (sockfd < 0) {
+    ret = -1;
+    goto ret_normal;
+  }
+
+  for (i = 0; i < hnums; ++i) {
+    srv = gethostbyname (hnames[i]);
+    if (srv != NULL) {
+      port = ports[i];
+      break;
+    }
+  }
+
+  if (srv == NULL) {
+    default_srv = gethostbyname (NTPUTIL_DEFAULT_HNAME);
+    if (default_srv == NULL) {
+      ret = -h_errno;
+      goto ret_close_sockfd;
+    }
+    srv = default_srv;
+    port = NTPUTIL_DEFAULT_PORT;
+  }
+
+  memset (&serv_addr, 0, sizeof (serv_addr));
+  serv_addr.sin_family = AF_INET;
+  memcpy ((uint8_t *) & serv_addr.sin_addr.s_addr,
+      (uint8_t *) srv->h_addr_list[0], (size_t) srv->h_length);
+  serv_addr.sin_port = htons (port);
+
+  ret = connect (sockfd, (struct sockaddr *) &serv_addr, sizeof (serv_addr));
+  if (ret < 0) {
+    ret = -errno;
+    goto ret_close_sockfd;
+  }
+
+  {
+    ntp_packet_t packet;
+    uint32_t recv_sec;
+    uint32_t recv_frac;
+    double frac;
+    ssize_t n;
+
+    memset (&packet, 0, sizeof (packet));
+
+    /* li = 0, vn = 3, mode = 3 */
+    packet.li_vn_mode = 0x1B;
+
+    /* Request */
+    n = write (sockfd, &packet, sizeof (packet));
+    if (n < 0) {
+      ret = -errno;
+      goto ret_close_sockfd;
+    }
+
+    pthread_spin_lock (&lock);
+    {
+      struct timespec ts;
+      err = clock_gettime (CLOCK_MONOTONIC, &ts);
+      local_monotonic_usec = ts.tv_sec * 1000000 + ts.tv_nsec / 1000;
+    }
+    pthread_spin_unlock (&lock);
+
+    /**
+     * Receive the packet from the server.
+     * @todo Use select() to wait for the response
+     */
+    n = read (sockfd, &packet, sizeof (packet));
+    if (n < 0) {
+      ret = -errno;
+      goto ret_close_sockfd;
+    }
+
+    /**
+     * @note ntp_timestamp_t recv_ts in ntp_packet_t means the timestamp as the packet
+     * left the NTP server. 'sec' corresponds to the seconds passed since 1900
+     * and 'frac' is needed to convert seconds to smaller units of a second
+     * such as microsceonds. Note that the bit/byte order of those data should
+     * be converted to the host's endianness.
+     */
+    recv_sec = ntohl (packet.xmit_ts.sec);
+    if (recv_sec == 0) {
+      ret = -1;
+      goto ret_close_sockfd;
+    }
+    recv_frac = ntohl (packet.xmit_ts.frac);
+
+    /**
+     * @note NTP uses an epoch of January 1, 1900 while the Unix epoch is
+     * the number of seconds that have elapsed since January 1, 1970. For this
+     * reason, we subtract 70 years worth of seconds from the seconds since 1900
+     */
+    ret = (int64_t) (recv_sec - NTPUTIL_TIMESTAMP_DELTA);
+    ret *= NTPUTIL_SEC_TO_USEC_MULTIPLIER;
```

Comment:
IMO, the overflow case is impossible since ```((uint32_t) recv_sec - 2208988800)  * 1000000``` is always smaller than ```9223372036854775807```, which is max of ```int64_t```.

Reply:
Ah, `recv_sec` is uint32.. Thanks for the explanation.